### PR TITLE
Bring OTA XML into closer alignment with the spec.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
1) Flag optional fields of QueryImage as optional.
2) Fix ordering of arguments for QueryImage.
3) Rename ApplyUpdateResponse to match the spec.
4) Use cluster-objects to send responses in example OTA provider.

#### Problem
Some optional fields are not optional

#### Change overview
See above.

#### Testing
Need to add tests that actually don't send the fields; not sure what the best setup for that is.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
I did some digging into the Linux codesize increase (or rather a similar one on Mac).  The main reason for the Mac one is that `CHIP_CONFIG_VERBOSE_VERIFY_OR_DIE` is true, so the `VerifyOrDie` in `Optional::Value` adds a bunch of code and string data for the logging call.  And it's separate code for every instantiation of `Optional`.  That's probably acceptable; any setup that really cares about codesize will not define `CHIP_CONFIG_VERBOSE_VERIFY_OR_DIE` to start with.

**Comment by andy31415** [comment]
@holbrookt @carol-apple @selissia ?

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp [hunk 1]
```diff
@@ -151,29 +149,26 @@ EmberAfStatus OTAProviderExample::HandleQueryImage(chip::app::CommandHandler * c
...
+    response.status                = queryStatus;
+    response.delayedActionTime     = mDelayedActionTimeSec;
+    response.imageURI              = chip::CharSpan(uriBuf, strlen(uriBuf));
+    response.softwareVersion       = newSoftwareVersion;
+    response.softwareVersionString = chip::CharSpan(kExampleSoftwareString, strlen(kExampleSoftwareString));
+    response.updateToken           = chip::ByteSpan(updateToken);
+    response.userConsentNeeded     = userConsentNeeded;
+    // TODO: Once our client is using APIs that handle optional arguments
+    // correctly, update QueryImageResponse to have the right things optional
+    // and just stop sending metadataForRequestor algether.
```
Comment (carol-apple) [inline examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:162]: I don't understand why we should stop sending `metadataForRequestor` when clients can handle optionals?

File: examples/ota-requestor-app/linux/main.cpp [hunk 2]
```diff
@@ -175,9 +175,8 @@ void OnConnected(void * context, OperationalDeviceProxy * operationalDeviceProxy
         ChipLogError(SoftwareUpdate, "Associate() failed: %" CHIP_ERROR_FORMAT, err.Format());
         return;
     }
-    err = cluster.QueryImage(successCallback, failureCallback, kExampleVendorId, kExampleProductId, kExampleHWVersion,
-                             kExampleSoftwareVersion, kExampleProtocolsSupported, exampleLocation, kExampleClientCanConsent,
-                             metadata);
+    err = cluster.QueryImage(successCallback, failureCallback, kExampleVendorId, kExampleProductId, kExampleSoftwareVersion,
```
Comment (carol-apple) [inline examples/ota-requestor-app/linux/main.cpp:178]: The order of the args here match that of the spec (👍 ) but the ordering in `OTAProviderExample` does not, e.g. `hardwareVersion` is listed after `productId` there.

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:162]: Well, right now we are sending an empty octet string; we could just not send it at all and save a few bytes on the wire....

**Review by carol-apple — APPROVED** [review]

**Review by holbrookt — APPROVED** [review]

**Review by jmeg-sfy — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**